File: dotnet/src/Envdoctor/Commands/SnapshotDiffCommand.cs
using Envdoctor.Core;
using Envdoctor.Models;
using Envdoctor.Runtime;
using Envdoctor.Utils;

namespace Envdoctor.Commands;

public sealed class SnapshotDiffArgs
{
    public string? Root { get; set; }
    public string A { get; set; } = "";
    public string B { get; set; } = "";
    public bool Json { get; set; }
}

public static class SnapshotDiffCommand
{
    /// Resolve a positional arg that may be a token string or a file path.
    private static RuntimeSnapshot LoadSnapshot(string root, string arg)
    {
        if (arg.Trim().StartsWith("envd1:", StringComparison.Ordinal))
            return Token.DecodeToken(arg);
        var file = Path.Combine(root, arg);
        if (!File.Exists(file))
            throw new SnapshotTokenException($"Not a snapshot token, and file not found: {arg}");
        return Token.ParseSnapshotJson(File.ReadAllText(file));
    }

    /// `envdoctor snapshot-diff <a> <b>` — compare two runtime snapshots.
    public static int Run(SnapshotDiffArgs args)
    {
        var root = Path.GetFullPath(args.Root ?? ".");

        RuntimeSnapshot a;
        RuntimeSnapshot b;
        try
        {
            a = LoadSnapshot(root, args.A);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"error {e.Message}");
            return ExitCodes.ExitUsage;
        }
        try
        {
            b = LoadSnapshot(root, args.B);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"error {e.Message}");
            return ExitCodes.ExitUsage;
        }

        var diff = Compare.CompareSnapshots(a, b);
        var exitCode = diff.Equivalent ? ExitCodes.ExitOk : ExitCodes.ExitIssues;

        if (args.Json)
        {
            Console.Out.WriteLine(DiffToJson(diff, exitCode));
            return exitCode;
        }

        RenderHuman(diff);
        return exitCode;
    }

    private static string DiffToJson(RuntimeDiff diff, int exitCode)
    {
        var globals = diff.Globals.Select(g =>
        {
            var obj = new JsonObject
            {
                { "ecosystem", g.Ecosystem },
                { "name", g.Name },
                { "status", g.Status.AsStr() },
            };
            if (g.A is not null)
                obj.Add("a", g.A);
            if (g.B is not null)
                obj.Add("b", g.B);
            return (object)obj;
        }).ToList();

        var tools = diff.Tools.Select(t =>
        {
            var obj = new JsonObject
            {
                { "name", t.Name },
                { "status", t.Status.AsStr() },
            };
            if (t.A is not null)
                obj.Add("a", t.A);
            if (t.B is not null)
                obj.Add("b", t.B);
            return (object)obj;
        }).ToList();

        // exitCode first, matching the reference CLI's key order.
        return Json.Pretty(new JsonObject
        {
            { "exitCode", exitCode },
            {
                "os", new JsonObject
                {
                    { "status", diff.Os.Status.AsStr() },
                    { "a", diff.Os.A },
                    { "b", diff.Os.B },
                }
            },
            { "tools", tools },
            { "pathReordered", diff.PathReordered },
            { "pathOnlyA", diff.PathOnlyA.Cast<object?>().ToList() },
            { "pathOnlyB", diff.PathOnlyB.Cast<object?>().ToList() },
            { "globals", globals },
            { "envFlagOnlyA", diff.EnvFlagOnlyA.Cast<object?>().ToList() },
            { "envFlagOnlyB", diff.EnvFlagOnlyB.Cast<object?>().ToList() },
            { "equivalent", diff.Equivalent },
        });
    }

    private static void RenderHuman(RuntimeDiff diff)
    {
        const string title = "RUNTIME DIFF";
        Console.Out.WriteLine(title);
        Console.Out.WriteLine(new string('─', title.Length * 2) + "\n");
        Console.Out.WriteLine("  A → B\n");

        if (diff.Os.Status == RuntimeStatus.Same)
            Console.Out.WriteLine($"  ✓ OS  {diff.Os.A}\n");
        else
            Console.Out.WriteLine($"  ⚠ OS  {diff.Os.A} → {diff.Os.B}\n");

        Console.Out.WriteLine("  Tools");
        foreach (var t in diff.Tools)
        {
            var a = t.A ?? "";
            var b = t.B ?? "";
            switch (t.Status)
            {
                case RuntimeStatus.Same:
                    Console.Out.WriteLine($"  ✓ {t.Name,-8} {a}");
                    break;
                case RuntimeStatus.Different:
                    Console.Out.WriteLine($"  ⚠ {t.Name,-8} {a} → {b}");
                    break;
                case RuntimeStatus.OnlyA:
                    Console.Out.WriteLine($"  ❌ {t.Name,-8} missing in B (A: {a})");
                    break;
                case RuntimeStatus.OnlyB:
                    Console.Out.WriteLine($"  ❌ {t.Name,-8} missing in A (B: {b})");
                    break;
            }
        }

        if (diff.PathReordered || diff.PathOnlyA.Count > 0 || diff.PathOnlyB.Count > 0)
        {
            Console.Out.WriteLine("\n  PATH");
            if (diff.PathReordered)
                Console.Out.WriteLine("  ⚠ same entries, different order");
            foreach (var p in diff.PathOnlyA)
                Console.Out.WriteLine($"  ❌ only in A: {p}");
            foreach (var p in diff.PathOnlyB)
                Console.Out.WriteLine($"  ❌ only in B: {p}");
        }

        if (diff.Globals.Count > 0)
        {
            Console.Out.WriteLine("\n  Globals");
            foreach (var g in diff.Globals)
            {
                var label = $"{g.Ecosystem}:{g.Name}";
                switch (g.Status)
                {
                    case RuntimeStatus.Different:
                        Console.Out.WriteLine($"  ⚠ {label}  {g.A ?? ""} → {g.B ?? ""}");
                        break;
                    case RuntimeStatus.OnlyA:
                        Console.Out.WriteLine($"  ❌ {label}  missing in B");
                        break;
                    case RuntimeStatus.OnlyB:
                        Console.Out.WriteLine($"  ❌ {label}  missing in A");
                        break;
                    case RuntimeStatus.Same:
                        break;
                }
            }
        }

        Console.Out.WriteLine(diff.Equivalent
            ? "\n  ✓ runtimes are equivalent"
            : "\n  ✗ runtime drift detected");
    }
}

File: swift/Sources/EnvdoctorKit/Commands/SnapshotDiffCommand.swift
import Foundation

public struct SnapshotDiffArgs {
    public var root: String?
    public var a: String
    public var b: String
    public var json = false

    public init(a: String, b: String) {
        self.a = a
        self.b = b
    }
}

public enum SnapshotDiffCommand {
    /// Resolve a positional arg that may be a token string or a file path.
    private static func loadSnapshot(_ root: String, _ arg: String) throws -> RuntimeSnapshot {
        if arg.trimmingCharacters(in: .whitespaces).hasPrefix("envd1:") {
            return try Token.decodeToken(arg)
        }
        let file = resolvePath(root, arg)
        guard FileManager.default.fileExists(atPath: file) else {
            throw Token.TokenError(description: "Not a snapshot token, and file not found: \(arg)")
        }
        let text = try String(contentsOfFile: file, encoding: .utf8)
        return try Token.parseSnapshotJson(text)
    }

    /// `envdoctor snapshot-diff <a> <b>` — compare two runtime snapshots.
    public static func run(_ args: SnapshotDiffArgs) throws -> Int {
        let root = resolveRoot(args.root)
        let a: RuntimeSnapshot
        let b: RuntimeSnapshot
        do {
            a = try loadSnapshot(root, args.a)
            b = try loadSnapshot(root, args.b)
        } catch let e as Token.TokenError {
            FileHandle.standardError.write("error \(e.description)\n")
            return 2
        } catch {
            FileHandle.standardError.write("error \(error.localizedDescription)\n")
            return 2
        }

        let diff = CompareSnapshots.compare(a, b)

        if args.json {
            print(Json.pretty(diffJsonValue(diff)))
            return diff.equivalent ? 0 : 1
        }

        renderHuman(diff)
        return diff.equivalent ? 0 : 1
    }

    /// `{exitCode, ...diff}` with `undefined` fields omitted, matching the
    /// reference's `JSON.stringify`.
    static func diffJsonValue(_ diff: RuntimeDiff) -> JSONValue {
        var root = JSONObject()
        root.set("exitCode", .int(diff.equivalent ? 0 : 1))

        var osObj = JSONObject()
        osObj.set("status", .string(diff.os.status.rawValue))
        osObj.set("a", .string(diff.os.a))
        osObj.set("b", .string(diff.os.b))
        root.set("os", .object(osObj))

        root.set("tools", .array(diff.tools.map { t in
            var tool = JSONObject()
            tool.set("name", .string(t.name))
            tool.set("status", .string(t.status.rawValue))
            if let a = t.a { tool.set("a", .string(a)) }
            if let b = t.b { tool.set("b", .string(b)) }
            return .object(tool)
        }))
        root.set("pathReordered", .bool(diff.pathReordered))
        root.set("pathOnlyA", .array(diff.pathOnlyA.map { .string($0) }))
        root.set("pathOnlyB", .array(diff.pathOnlyB.map { .string($0) }))
        root.set("globals", .array(diff.globals.map { g in
            var global = JSONObject()
            global.set("ecosystem", .string(g.ecosystem))
            global.set("name", .string(g.name))
            global.set("status", .string(g.status.rawValue))
            if let a = g.a { global.set("a", .string(a)) }
            if let b = g.b { global.set("b", .string(b)) }
            return .object(global)
        }))
        root.set("envFlagOnlyA", .array(diff.envFlagOnlyA.map { .string($0) }))
        root.set("envFlagOnlyB", .array(diff.envFlagOnlyB.map { .string($0) }))
        root.set("equivalent", .bool(diff.equivalent))
        return .object(root)
    }

    private static func renderHuman(_ diff: RuntimeDiff) {
        let title = "RUNTIME DIFF"
        print(title)
        print(rule(title.count * 2) + "\n")
        print("  A → B\n")

        if diff.os.status == .same {
            print("  ✓ OS  \(diff.os.a)\n")
        } else {
            print("  ⚠ OS  \(diff.os.a) → \(diff.os.b)\n")
        }

        print("  Tools")
        for t in diff.tools {
            switch t.status {
            case .same:
                print("  ✓ \(padEnd(t.name, 8)) \(t.a ?? "")")
            case .different:
                print("  ⚠ \(padEnd(t.name, 8)) \(t.a ?? "") → \(t.b ?? "")")
            case .onlyA:
                print("  ❌ \(padEnd(t.name, 8)) missing in B (A: \(t.a ?? ""))")
            case .onlyB:
                print("  ❌ \(padEnd(t.name, 8)) missing in A (B: \(t.b ?? ""))")
            }
        }

        if diff.pathReordered || !diff.pathOnlyA.isEmpty || !diff.pathOnlyB.isEmpty {
            print("\n  PATH")
            if diff.pathReordered {
                print("  ⚠ same entries, different order")
            }
            for p in diff.pathOnlyA {
                print("  ❌ only in A: \(p)")
            }
            for p in diff.pathOnlyB {
                print("  ❌ only in B: \(p)")
            }
        }

        if !diff.globals.isEmpty {
            print("\n  Globals")
            for g in diff.globals {
                let label = "\(g.ecosystem):\(g.name)"
                switch g.status {
                case .different:
                    print("  ⚠ \(label)  \(g.a ?? "") → \(g.b ?? "")")
                case .onlyA:
                    print("  ❌ \(label)  missing in B")
                case .onlyB:
                    print("  ❌ \(label)  missing in A")
                case .same:
                    break
                }
            }
        }

        let status = diff.equivalent ? "✓ runtimes are equivalent" : "✗ runtime drift detected"
        print("\n  \(status)")
    }
}
